### PR TITLE
Rename `test_cache_is_invalidated` and remove `assert study._thread_local.cached_all_trials is None`

### PR DESCRIPTION
**Pull request by nabenabe0928**
## Motivation

The original version test checks the cache clear after before_trial, so if we call `_get_all_trials` with `use_cache=True` at `before_trial`, this unit test fails.
To avoid this problem, we actually need to check the cache clear before `before_trial`.
To this end, I introduced a mock function to do so.

## Description of the changes

- Add a mock function so that we can check the cache clear before each trial

> [!NOTE]
> I thought through if I should add any tests for this regard to `study_tests`, but as this check should happen before `before_trial`, which is a method of each sampler, I simply modified the existing test in `test_samplers.py` instead of `study_tests`.

### CODE REVIEW

**Comment by nabenabe0928** [comment]
@y0z 
Could you review this PR?

File: tests/samplers_tests/test_samplers.py
Comment (y0z) [inline tests/samplers_tests/test_samplers.py:1057]: This line causes the "tests/samplers_tests/test_samplers.py:1057: error: Cannot assign to a method  [method-assign]" mypy error.

I think [`mock.patch.object`](https://docs.python.org/3.13/library/unittest.mock.html) is a better way to replace a class method.

Example:
```python
import unittest.mock


class C:
    def method(self):
        print("hello")

def mock_method():
    print("hi")

c = C()

c.method()
with unittest.mock.patch.object(C, 'method', unittest.mock.Mock(side_effect=mock_method)):
    c.method()
```
Output:
```
hello
hi
```

**Review by y0z — APPROVED** [review]
LGTM

**Comment by nabenabe0928** [comment]
@HideakiImamura 
Could you review this PR?

**Comment by HideakiImamura** [comment]
Thanks for the PR. IMO, the line of `assert study._thread_local.cached_all_trials is None` should be removed from the current test function, because the it is not a specification that the trials are not cached before the call of suggest methods at now. 

If we want to test the behavior of the trial cache between sampler's before trial method and the cache creation, it would be better to add the additional test cases. However, I think we don't have to specify this behavior, so removing the assertion is sufficient, IMO.

**Review by HideakiImamura — APPROVED** [review]
Thanks for the update. Looks very good.

**Comment by HideakiImamura** [comment]
@y0z Please merge this PR after your approval, since the direction of this PR is completely changed.

**Comment by y0z** [comment]
The change looks OK. I rename the PR title since the first title is no longer proper.

**Pull request merged**
